hazelcast/src/main/java/com/hazelcast/internal/probing/ProbeRegistryImpl.java

```diff
@@ -0,0 +1,681 @@
+package com.hazelcast.internal.probing;
+
+import static com.hazelcast.internal.probing.CharSequenceUtils.appendEscaped;
+import static com.hazelcast.internal.probing.CharSequenceUtils.appendUnescaped;
+import static com.hazelcast.util.StringUtil.getterIntoProperty;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+import com.hazelcast.internal.metrics.Probe;
+import com.hazelcast.internal.metrics.ProbeLevel;
+import com.hazelcast.internal.probing.ProbingCycle.Tagging;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.util.Clock;
+
+public final class ProbeRegistryImpl implements ProbeRegistry {
+
+    private static final ILogger LOGGER = Logger.getLogger(ProbeRegistryImpl.class);
+
+    private static final Object[] EMPTY_ARGS = new Object[0];
+
+    private static final Map<Class<?>, ProbeAnnotatedType> PROBE_METADATA = 
+            new ConcurrentHashMap<Class<?>, ProbeAnnotatedType>();
+
+    private final Deque<ProbeSourceEntry> sources = new ConcurrentLinkedDeque<ProbeSourceEntry>();
+
+    @Override
+    public void register(ProbeSource source) {
+        for (ProbeSourceEntry e : sources) {
+            if (e.source.equals(source)) {
+                LOGGER.info("Probe source tried to register more then once: "
+                        + source.getClass().getSimpleName());
+                return; // avoid adding the very same instance more then once
+            }
+        }
+        sources.add(new ProbeSourceEntry(source));
+    }
+
+    @Override
+    public void registerIfSource(Object source) {
+        if (source instanceof ProbeSource) {
+            register((ProbeSource) source);
+        }
+    }
+
+    @Override
+    public ProbeRenderContext newRenderingContext() {
+        return new ProbingCycleImpl(sources);
+    }
+
+    @Override
+    public ProbeRenderContext newRenderingContext(Class<? extends ProbeSource>... filter) {
+        Set<ProbeSourceEntry> filtered = new HashSet<ProbeSourceEntry>();
+        for (ProbeSourceEntry e : sources) {
+            for (Class<?> accepted : filter) {
+                if (accepted == e.source.getClass()) {
+                    filtered.add(e);
+                }
+            }
+        }
+        return new ProbingCycleImpl(filtered);
+    }
+
+    /**
+     * Wrapper for a {@link ProbeSource} for the state to keep track of it potential
+     * need to be updated before being probed.
+     */
+    private static final class ProbeSourceEntry {
+
+        final ProbeSource source;
```

Comment:
Why are these fields not private? Same goes for constructor. If it is for testing purposes; please add a line of comment.

Reply:
Its a private class. Fields are private already. Marking the fields private explicitly changes nothing when it comes to accessibility so its just litter that I avoid.